pyomo/contrib/solver/solvers/knitro/engine.py

```diff
@@ -0,0 +1,422 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2025
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from collections.abc import Callable, Iterable, Mapping, MutableMapping, Sequence
+from types import MappingProxyType
+from typing import Any, Optional, TypeVar
+
+from pyomo.common.enums import ObjectiveSense
+from pyomo.common.errors import DeveloperError
+from pyomo.common.numeric_types import value
+from pyomo.contrib.solver.solvers.knitro.api import knitro
+from pyomo.contrib.solver.solvers.knitro.callback import build_callback_handler
+from pyomo.contrib.solver.solvers.knitro.package import Package
+from pyomo.contrib.solver.solvers.knitro.typing import (
+    BoundType,
+    Callback,
+    ItemData,
+    ItemType,
+    StructureType,
+    ValueType,
+)
+from pyomo.contrib.solver.solvers.knitro.utils import NonlinearExpressionData
+from pyomo.core.base.constraint import ConstraintData
+from pyomo.core.base.objective import ObjectiveData
+from pyomo.core.base.var import VarData
+from pyomo.repn.standard_repn import generate_standard_repn
+
+
+def parse_bounds(
+    items: Iterable[ItemType], idx_map: Mapping[int, int]
+) -> Mapping[BoundType, MutableMapping[int, float]]:
+    bounds_map = {bnd_type: {} for bnd_type in BoundType}
+    for item in items:
+        i = idx_map[id(item)]
+        if isinstance(item, VarData):
+            if item.fixed:
+                bounds_map[BoundType.EQ][i] = value(item.value)
+                continue
+            if item.has_lb():
+                bounds_map[BoundType.LO][i] = value(item.lb)
+            if item.has_ub():
+                bounds_map[BoundType.UP][i] = value(item.ub)
```

Comment:
It would be more efficient to do:
```suggestion
            lb, ub = item.bounds
            if lb is not None:
                bounds_map[BoundType.LO][i] = lb
            if ub is not None:
                bounds_map[BoundType.UP][i] = ub
```